### PR TITLE
(maint) Deprecate settings used by `puppet master`

### PR DESCRIPTION
**Pull request by justinstoller**
As part of PUP-8591 we removed the master application in Puppet 6. This
orphaned many settings that are not applicable to running a master with
Puppet Server or are better served living within its configuration
files.

To prepare for the removal of the puppet master application in Puppet 6,
this patch deprecates those settings.

---

Also correctly deprecates configprint.

### CODE REVIEW

File: lib/puppet/defaults.rb [hunk 1]
```diff
@@ -257,8 +257,9 @@ def self.valid_file_checksum_types
           on the CLI.",
     },
     :configprint => {
-        :default  => "",
-        :desc     => "Prints the value of a specific configuration setting.  If the name of a
+        :default    => "",
+        :deprecated => :completely,
+        :desc       => "Prints the value of a specific configuration setting.  If the name of a
```
Comment (justinstoller) [inline lib/puppet/defaults.rb]: Ugh, I caught this one while skimming the file and added it. I suppose it should be in a different PR, or at least different commit though?

**Comment by justinstoller** [comment]
Broke out the configprint deprecation improvement into its own commit and updated the commit message of the master application deprecations to better follow our guidelines.

> Are default_manifest and disable_per_environment_manifest still used by puppetserver?

They're used in `Puppet::Node::Environment` my assumption was that they were still used by the compiler when not integrated with an ENC.

File: lib/puppet/defaults.rb [hunk 2]
```diff
@@ -1290,13 +1298,16 @@ def self.valid_file_checksum_types
     :rest_authconfig => {
       :default    => "$confdir/auth.conf",
       :type       => :file,
+      :deprecated  => :completely,
       :desc       => "The configuration file that defines the rights to the different
-      rest indirections.  This can be used as a fine-grained
-      authorization system for `puppet master`.",
+      rest indirections.  This can be used as a fine-grained authorization system for
+      `puppet master`.  The `puppet master` command is deprecated and Puppet Server
+      uses its own auth.conf that must be placed within its configuration directory.",
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: Will puppetserver still use `:rest_authconf` if `use-legacy-auth-conf` is enabled?

File: lib/puppet/defaults.rb (hunk 1 above)
Comment (joshcooper) [reply lib/puppet/defaults.rb]: The `--configprint` option/setting applies to all subcommands, e.g.

```
$ bundle exec puppet apply --configprint server
localhost
```

But it is deprecated (with `puppet config` being the replacement), and I'm fine deprecating `configprint` in this PR.

File: lib/puppet/defaults.rb [hunk 3]
```diff
@@ -1209,11 +1211,17 @@ def self.valid_file_checksum_types
   define_settings(:master,
     :user => {
       :default    => "puppet",
-      :desc       => "The user puppet master should run as.",
+      :desc       => "The user puppet master should run as. Unused,
+      Puppet Server utilizes a system's service framework to run as
+      this user.",
+      :deprecated  => :completely,
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: I'm don't think we can deprecate `user` and `group` quite yet, because the host's cert/key (not the CA stuff) is still shared between the `puppet-agent` and `puppet-server` packages. So puppet needs to know which `user` and `group` to map the `service` account to, so that puppetserver can read those files when running as `puppet` or `pe-puppet`.

File: lib/puppet/defaults.rb [hunk 4]
```diff
@@ -1259,16 +1267,16 @@ def self.valid_file_checksum_types
...
         running under a Rack server like Passenger, that web server will have
         its own logging behavior."
     },
     :masterport => {
       :default    => 8140,
-      :desc       => "The port for puppet master traffic. For puppet master,
-      this is the port to listen on; for puppet agent, this is the port
-      to make requests on. Both applications use this setting to get the port.",
+      :desc       => "The port for puppet puppet agent to make requests on.
+      Only used by the puppet agent.",
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: It's mainly puppet agent, but it also affects puppet subcommand making REST requests to the puppetserver, e.g. `puppet facts upload`. Not sure how best to word that... maybe something related to the `/puppet` route? Also it's more like the default port, as it can be overridden for specific endpoints, e.g. `report_port`.

File: lib/puppet/test/test_helper.rb [hunk 5]
```diff
@@ -221,9 +221,6 @@ def self.initialize_settings_before_each()
       # Initialize "app defaults" settings to a good set of test values
       Puppet.settings.initialize_app_defaults(app_defaults_for_tests)
 
-      # Avoid opening ports to the outside world
-      Puppet.settings[:bindaddress] = "127.0.0.1"
-
```
[the remark this answers is not in the document]
Comment (justinstoller) [reply lib/puppet/test/test_helper.rb]: Good question, it passed on my machine (:  I'll update the master application spec directly to avoid deprecation notices in all the specs.

File: lib/puppet/defaults.rb (hunk 3 above)
Comment (justinstoller) [reply lib/puppet/defaults.rb]: Grumble... Good catch, with our recent work with `puppetserver ca create` I wasn't thinking about the use case of using an agent to bootstrap a compile master's PKI.

You mind if I update the description and hook to disallow values besides `puppet` and `pe-puppet` and call out that, if set, this must be the value that matches their kind of installation (commercial or open source)?

File: lib/puppet/defaults.rb (hunk 2 above)
Comment (justinstoller) [reply lib/puppet/defaults.rb]: Yes, it will for routes that are still implemented using Puppet's Ruby rest interface. That excludes the CA and admin APIs which are in clojure (and anything else we move to clojure). In Puppet Server 5 `use-legacy-auth-conf` is [off by default](https://puppet.com/docs/puppetserver/5.3/config_file_auth.html#aside-changes-to-authorization-in-puppet-server-500)(docs) and [warns if enabled](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj#L279-L281)(code).

File: lib/puppet/defaults.rb (hunk 3 above)
Comment (joshcooper) [reply lib/puppet/defaults.rb]: I think I'd prefer to leave out the deprecation. The `user` and `group` settings are typically used to share ssl files, but it's also used to share `server_datadir`, `report_dir`, ... While puppetserver has a separate `vardir` by default, it's configurable, so I'm not sure what workflows would be affected.
